apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRDASource.java

```diff
@@ -0,0 +1,152 @@
+            .build(),
+        callerFactory,
+        Clock.systemDefaultZone(),
+        appMetrics);
+  }
+
+  @VisibleForTesting
+  GrpcRDASource(
+      ManagedChannel channel,
+      GrpcStreamCaller.Factory<T> callerFactory,
+      Clock clock,
+      MetricRegistry appMetrics) {
+    this.callerFactory = callerFactory;
+    this.channel = channel;
+    this.clock = clock;
+    callsMeter = appMetrics.meter(CALLS_METER);
+    recordsReceivedMeter = appMetrics.meter(RECORDS_RECEIVED_METER);
+    recordsStoredMeter = appMetrics.meter(RECORDS_STORED_METER);
+    batchesMeter = appMetrics.meter(BATCHES_METER);
+  }
+
+  @Override
+  public int retrieveAndProcessObjects(
+      int maxToProcess, int maxPerBatch, Duration maxRunTime, RDASink<PreAdjudicatedClaim> sink)
+      throws ProcessingException {
+    callsMeter.mark();
+    int processed = 0;
+    try {
+      final GrpcStreamCaller<T> caller = callerFactory.createCaller(channel);
+      final List<PreAdjudicatedClaim> batch = new ArrayList<>();
+      Instant now = clock.instant();
+      final Instant stopTime = now.plus(maxRunTime);
+      while (shouldContinue(processed, maxToProcess, now, stopTime)) {
+        final Iterator<T> resultIterator = caller.callService(Duration.between(now, stopTime));
+        now = clock.instant();
+        while (resultIterator.hasNext() && shouldContinue(processed, maxToProcess, now, stopTime)) {
+          final T result = resultIterator.next();
+          final PreAdjudicatedClaim claim = caller.convertResultToClaim(result);
+          recordsReceivedMeter.mark();
+          batch.add(claim);
+          if (batch.size() >= maxPerBatch) {
+            processed += submitBatchToSink(sink, batch);
+          }
+          now = clock.instant();
+        }
+        if (batch.size() > 0) {
+          processed += submitBatchToSink(sink, batch);
+        }
+      }
+    } catch (ProcessingException ex) {
+      throw new ProcessingException(ex.getCause(), processed + ex.getProcessedCount());
+    } catch (Exception ex) {
+      throw new ProcessingException(ex, processed);
+    }
+    return processed;
+  }
+
+  @Override
+  public void close() throws Exception {
+    if (channel != null) {
+      channel.shutdownNow().awaitTermination(5, TimeUnit.SECONDS);
+      channel = null;
+    }
+  }
+
+  private boolean shouldContinue(int processed, int maxToProcess, Instant now, Instant stopTime) {
+    if (processed >= maxToProcess) {
+      LOGGER.info("exiting loop after processing max number of records: processed={}", processed);
+      return false;
+    }
+    if (now.compareTo(stopTime) >= 0) {
+      LOGGER.info("exiting loop after reaching max runtime");
+      return false;
+    }
+    return true;
+  }
+
+  private int submitBatchToSink(RDASink<PreAdjudicatedClaim> sink, List<PreAdjudicatedClaim> batch)
+      throws ProcessingException {
+    LOGGER.info("submitting batch to sink: size={}", batch.size());
+    int processed = sink.writeBatch(batch);
```

Comment:
would it be useful to measure inserted vs updated at this level or do you think we should log those details in the sink?

Reply:
I would prefer to do that in the sink since that is where that particular determination would be made.  Another possibility in the sink would be records rejected or ignored for some reason.

Reply:
Rejections are probably something we're going to want to complain loudly about and possibly even halt processing for; thats typically been the case for the existing RIF pipeline job, but we can cross that bridge when we start implementing the sink